sys/kern/tty.c

```diff
@@ -173,6 +216,123 @@ static void tty_outq_putc(tty_t *tty, uint8_t c) {
   }
 }
 
+/*
+ * Echo a typed character to the terminal.
+ * Call with `tty->t_lock` held.
+ */
+static void tty_echo(tty_t *tty, uint8_t c) {
+  assert(mtx_owned(&tty->t_lock));
+
+  int lflag = tty->t_lflag;
+  /* If ECHO is not set, only echo newlines provided ECHONL is set. */
+  if (!(lflag & ECHO) && (!(lflag & ECHONL) || c != '\n'))
+    return;
+  /* If ECHOCTL is set, echo control characters as ^A, ^B etc. */
+  if ((lflag & ECHO) && ((c <= 0x1f && c != '\t' && c != '\n') || c == 0x7f)) {
```

Comment:
Two magic constants... is there any macro that defines them?

Reply:
I have added one.